### ISSUE TITLE
"unexpected EOF" with Docker 1.10-rc1 and proxy --rewrite-inspect

### ISSUE THREAD

**Issue by errordeveloper** [weaveworks/weave#1911]
```
> docker -H unix:///var/run/docker.sock exec weaveproxy ls / > /dev/null
> docker -H unix:///var/run/weave/weave.sock exec weaveproxy ls / > /dev/null
unexpected EOF
```

**Comment by errordeveloper** [comment]
Also:

```
docker@kubedev:~$ unset DOCKER_HOST
docker@kubedev:~$ weave status

        Version: 1.4.2

        Service: router
       Protocol: weave 1..2
           Name: be:6a:0d:46:d3:f2(kubedev)
     Encryption: disabled
  PeerDiscovery: enabled
        Targets: 0
    Connections: 0
          Peers: 1
 TrustedSubnets: none

        Service: ipam
         Status: ready
          Range: 10.32.0.0-10.47.255.255
  DefaultSubnet: 10.32.0.0/12

        Service: dns
         Domain: weave.local.
       Upstream: 192.168.131.2
            TTL: 1
        Entries: 1

        Service: proxy
        Address: unix:///var/run/weave/weave.sock

docker@kubedev:~$ eval `weave env`
docker@kubedev:~$ weave status

        Version: 1.4.2

        Service: router
       Protocol: weave 1..2
           Name: be:6a:0d:46:d3:f2(kubedev)
     Encryption: disabled
  PeerDiscovery: enabled
        Targets: 0
    Connections: 0
          Peers: 1
 TrustedSubnets: none

        Service: ipam
         Status: ready
          Range: 10.32.0.0-10.47.255.255
  DefaultSubnet: 10.32.0.0/12

        Service: dns
         Domain: weave.local.
       Upstream: 192.168.131.2
            TTL: 1
        Entries: 1

        Service: proxy
        Address: time="2016-01-19T16:24:23Z" level=fatal msg="unexpected EOF" 
unix:///var/run/weave/weave.sock

docker@kubedev:~$ 
```

**Comment by errordeveloper** [comment]
also, somewhat weird (and not test with non-busybox version of shell yet), but still seeming related due to the appearance of `time=...`:
```
docker@kubedev:/var/run$ unset DOCKER_HOST
docker@kubedev:/var/run$ eval `weave env
> `
docker@kubedev:/var/run$ echo $DOCKER_HOST
unix:///var/run/weave/weave.sock
docker@kubedev:/var/run$ unset DOCKER_HOST
docker@kubedev:/var/run$ eval `weave env`
docker@kubedev:/var/run$ echo $DOCKER_HOST
unix:///var/run/weave/weave.sock
docker@kubedev:/var/run$ eval `weave env
> `
-sh: export: unix:///var/run/weave/weave.sock: bad variable name
docker@kubedev:/var/run$ echo $DOCKER_HOST
time=2016-01-19T10:23:34Z
docker@kubedev:/var/run$ 
```

**Comment by errordeveloper** [comment]
This time string might appear do to this being a development build or and some debug logging is enabled something, but I'm not sure. I'm using the Boot2Docker 1.10-rc1 ISO (`https://github.com/tianon/boot2docker-legacy/releases/download/v1.10.0-rc1/boot2docker.iso`).

**Comment by errordeveloper** [comment]
```
docker@kubedev:~$ unset DOCKER_HOST
docker@kubedev:~$ eval $(weave env)
docker@kubedev:~$ weave env
export DOCKER_HOST=time="2016-01-19T16:51:44Z"
unix:///var/run/weave/weave.sock ORIG_DOCKER_HOST=unix:///var/run/weave/weave.sock
docker@kubedev:~$ WEAVE_DEBUG=1 weave env
+ SCRIPT_VERSION=1.4.2
+ [ 1.4.2 = (unreleased version) ]
+ IMAGE_VERSION=1.4.2
+ IMAGE_VERSION=1.4.2
+ MIN_DOCKER_VERSION=1.6.0
+ DOCKERHUB_USER=weaveworks
+ BASE_EXEC_IMAGE=weaveworks/weaveexec
+ EXEC_IMAGE=weaveworks/weaveexec:1.4.2
+ cut -s -d: -f1
+ echo unix:///var/run/weave/weave.sock
+ PROXY_HOST=unix
+ PROXY_HOST=unix
+ DOCKER_CLIENT_HOST=unix:///var/run/weave/weave.sock
+ IP_REGEXP=[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}
+ CIDR_REGEXP=[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}/[0-9]{1,2}
+ check_docker_version
+ sed -n -e s|^Docker version \([0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*\).*|\1|p
+ docker -v
+ DOCKER_VERSION=1.10.0
+ [ -z 1.10.0 ]
+ version_lt 1.10.0 1.6.0
+ VERSION_MAJOR=1
+ REST=1.10 VERSION_MINOR=10
+ VERSION_PATCH=0
+ MIN_VERSION_MAJOR=1
+ REST=1.6 MIN_VERSION_MINOR=6
+ MIN_VERSION_PATCH=0
+ [ ( 1 -lt 1 ) -o ( 1 -eq 1 -a ( 10 -lt 6 -o ( 10 -eq 6 -a ( 0 -lt 0 ) ) ) ) ]
+ return 1
+ [ env = --local ]
+ [ env = --help -o env = help ]
+ [ env = version -a -z  ]
+ [ env = env -a  = --restore ]
+ [ env = run -a -z  ]
+ [ -z  ]
+ exec_remote env
+ docker_sock_options
+ grep -q ^unix://
+ echo unix:///var/run/weave/weave.sock
+ echo -v /var/run/weave/weave.sock:/var/run/weave/weave.sock -e DOCKER_HOST
+ docker run --rm --privileged --net=host -v /var/run/weave/weave.sock:/var/run/weave/weave.sock -e DOCKER_HOST -v /proc:/hostproc -e PROCFS=/hostproc -e DOCKERHUB_USER=weaveworks -e WEAVE_VERSION -e WEAVE_DEBUG -e WEAVE_DOCKER_ARGS -e WEAVEPROXY_DOCKER_ARGS -e WEAVE_PASSWORD -e WEAVE_PORT -e WEAVE_CONTAINER_NAME -e WEAVE_MTU -e WEAVE_NO_FASTDP -e WEAVE_NO_BRIDGED_FASTDP -e WEAVE_NO_PLUGIN -e DOCKER_BRIDGE -e DOCKER_CLIENT_HOST=unix:///var/run/weave/weave.sock -e DOCKER_CLIENT_ARGS -e PROXY_HOST=unix -e COVERAGE weaveworks/weaveexec:1.4.2 --local env
+ SCRIPT_VERSION=1.4.2
+ [ 1.4.2 = (unreleased version) ]
+ IMAGE_VERSION=1.4.2
+ IMAGE_VERSION=1.4.2
+ MIN_DOCKER_VERSION=1.6.0
+ DOCKERHUB_USER=weaveworks
+ BASE_EXEC_IMAGE=weaveworks/weaveexec
+ EXEC_IMAGE=weaveworks/weaveexec:1.4.2
+ PROXY_HOST=unix
+ PROXY_HOST=unix
+ DOCKER_CLIENT_HOST=unix:///var/run/weave/weave.sock
+ IP_REGEXP=[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}
+ CIDR_REGEXP=[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}/[0-9]{1,2}
+ check_docker_version
+ sed -n -e s|^Docker version \([0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*\).*|\1|p
+ docker -v
+ DOCKER_VERSION=1.6.2
+ [ -z 1.6.2 ]
+ version_lt 1.6.2 1.6.0
+ VERSION_MAJOR=1
+ REST=1.6 VERSION_MINOR=6
+ VERSION_PATCH=2
+ MIN_VERSION_MAJOR=1
+ REST=1.6 MIN_VERSION_MINOR=6
+ MIN_VERSION_PATCH=0
+ [ ( 1 -lt 1 ) -o ( 1 -eq 1 -a ( 6 -lt 6 -o ( 6 -eq 6 -a ( 2 -lt 0 ) ) ) ) ]
+ return 1
+ [ --local = --local ]
+ shift 1
+ IS_LOCAL=1
+ [ env = --help -o env = help ]
+ [ env = version -a -z 1 ]
+ [ env = env -a  = --restore ]
+ [ env = run -a -z 1 ]
+ [ -z 1 ]
+ BASE_IMAGE=weaveworks/weave
+ IMAGE=weaveworks/weave:1.4.2
+ BASE_PLUGIN_IMAGE=weaveworks/plugin
+ PLUGIN_IMAGE=weaveworks/plugin:1.4.2
+ PLUGIN_CONTAINER_NAME=weaveplugin
+ PROCFS=/hostproc
+ DOCKER_BRIDGE=docker0
+ CONTAINER_NAME=weave
+ BRIDGE=weave
+ DATAPATH=datapath
+ BRIDGE_IFNAME=vethwe-bridge
+ DATAPATH_IFNAME=vethwe-datapath
+ CONTAINER_IFNAME=ethwe
+ PORT=6783
+ HTTP_PORT=6784
+ PROXY_PORT=12375
+ PROXY_CONTAINER_NAME=weaveproxy
+ COVERAGE_ARGS=
+ [ -n  ]
+ id -u
+ [ 0 = 0 ]
+ read sys maj min
+ sed -n -e s|^\([^ ]*\) \([0-9][0-9]*\)\.\([0-9][0-9]*\).*|\1 \2 \3|p
+ uname -s -r
+ [ Linux != Linux ]
+ [ ( 4 -eq 3 -a 1 -ge 8 ) -o 4 -gt 3 ]
+ command_exists ip
+ command -v ip
+ ip netns list
+ command_exists nsenter
+ command -v nsenter
+ [ 1 -gt 0 ]
+ COMMAND=env
+ shift 1
+ [ env = env ]
+ proxy_addr
+ proxy_addrs
+ http_call_unix weaveproxy status.sock GET /status
+ addr=+ container=weaveproxy
+ socket=status.sock
+ http_verb=GET
+ url=/status
+ shift 4
+ docker exec weaveproxy curl -s -S -X GET --unix-socket status.sock http:/status
time="2016-01-19T16:51:57Z" level=fatal msg="unexpected EOF" 
+ output=unix:///var/run/weave/weave.sock
+ [ -n unix:///var/run/weave/weave.sock ]
+ echo unix:///var/run/weave/weave.sock
unix:///var/run/weave/weave.sock
+ sed s/0.0.0.0/unix/g
+ echo + container=weaveproxy
+ socket=status.sock
+ http_verb=GET
+ url=/status
+ shift 4
+ docker exec weaveproxy curl -s -S -X GET --unix-socket status.sock http:/status
time="2016-01-19T16:51:57Z" level=fatal msg="unexpected EOF" 
+ output=unix:///var/run/weave/weave.sock
+ [ -n unix:///var/run/weave/weave.sock ]
+ echo unix:///var/run/weave/weave.sock
unix:///var/run/weave/weave.sock
+ addr=+ container=weaveproxy
+ socket=status.sock
+ http_verb=GET
+ url=/status
+ shift 4
+ docker exec weaveproxy curl -s -S -X GET --unix-socket status.sock http:/status
time="2016-01-19T16:51:57Z" level=fatal msg="unexpected EOF" 
+ output=unix:///var/run/weave/weave.sock
+ [ -n unix:///var/run/weave/weave.sock ]
+ echo unix:///var/run/weave/weave.sock
unix:///var/run/weave/weave.sock
+ cut -d   -f1
+ echo + container=weaveproxy
+ socket=status.sock
+ http_verb=GET
+ url=/status
+ shift 4
+ docker exec weaveproxy curl -s -S -X GET --unix-socket status.sock http:/status
time="2016-01-19T16:51:57Z" level=fatal msg="unexpected EOF" 
+ output=unix:///var/run/weave/weave.sock
+ [ -n unix:///var/run/weave/weave.sock ]
+ echo unix:///var/run/weave/weave.sock
unix:///var/run/weave/weave.sock
+ PROXY_ADDR=+
+
+
+
+
+
time="2016-01-19T16:51:57Z"
+
+
+
unix:///var/run/weave/weave.sock
+ [ +
+
+
+
+
+
time="2016-01-19T16:51:57Z"
+
+
+
unix:///var/run/weave/weave.sock = unix:///var/run/weave/weave.sock ]
+ RESTORE=ORIG_DOCKER_HOST=unix:///var/run/weave/weave.sock
+ echo export DOCKER_HOST=+
+
+
+
+
+
time="2016-01-19T16:51:57Z"
+
+
+
unix:///var/run/weave/weave.sock ORIG_DOCKER_HOST=unix:///var/run/weave/weave.sock
export DOCKER_HOST=+
+
+
+
+
+
time="2016-01-19T16:51:57Z"
+
+
+
unix:///var/run/weave/weave.sock ORIG_DOCKER_HOST=unix:///var/run/weave/weave.sock
+ exit 0
docker@kubedev:~$ 
```

**Comment by rade** [comment]
The obvious bug here is
```
proxy_addrs() {
    if addr="$(http_call_unix $PROXY_CONTAINER_NAME status.sock GET /status 2>&1)" ; then
...
```
which should have a `2>/dev/null` instead of `2>&1`.

That still leaves the question though what is producing that stderr output. I suspect it's the weaveproxy, since the erroneous behaviour only occurs when `weave env` is run against the proxy.

**Comment by errordeveloper** [comment]
I think basically all Docker wants to say is `unexpected EOF`, but when it's a ran from a non-interactive shell, it formats it with as a log message (i.e. `time="2016-01-19T16:24:23Z" level=fatal msg="unexpected EOF").

**Comment by errordeveloper** [comment]
Turns out that this can only be reproduced when the proxy is launched with `--rewrite-inspect` flag.

**Comment by awh** [comment]
Current suspicion is that the exec inspect interceptor is to blame. Pre and post responses with 1.10-rc1:
```
{
  "ID": "a1034ed78650a4308ec0fbff6bb54454e03d9e0305273e02d332ba952e84df7c",
  "Running": false,
  "ExitCode": 0,
  "ProcessConfig": {
    "tty": false,
    "entrypoint": "ls",
    "arguments": [],
    "privileged": false,
    "user": ""
  },
  "OpenStdin": false,
  "OpenStderr": true,
  "OpenStdout": true,
  "CanRemove": false,
  "ContainerID": "eb4900dc985bb57a265739d8524acf4128489af4fba26d742f21cf5be16d635c",
  "DetachKeys": ""
}
```
```
{
  "CanRemove": false,
  "Container": {},
  "ContainerID": "eb4900dc985bb57a265739d8524acf4128489af4fba26d742f21cf5be16d635c",
  "DetachKeys": "",
  "ExitCode": 0,
  "ID": "a1034ed78650a4308ec0fbff6bb54454e03d9e0305273e02d332ba952e84df7c",
  "OpenStderr": true,
  "OpenStdin": false,
  "OpenStdout": true,
  "ProcessConfig": {
    "arguments": [],
    "entrypoint": "ls",
    "privileged": false,
    "tty": false,
    "user": ""
  },
  "Running": false
}
```
We're inserting an empty `Container:{}` node when exec'ing against a container with no weave IPs, this seems to have started causing a problem for some reason. It's a bit odd because both the Docker 1.10-rc1 and Docker 1.6.2 client in the weaveexec container are affected.

**Comment by awh** [comment]
I think I have found the problem - in previous versions of Docker, the exec inspect JSON looks like the full container inspect, including a complete `Containers: { ... }` subtree. It would seem that they have thinned things down to be relevant to `exec` only, and we are now inserting an empty `Containers: {}` node which is breaking both 1.6.2 and 1.10 clients (and presumably others besides). Will modify interceptor to test for presence of `Containers: {}` node and skip rewrite if it's not present.

**Comment by bboreham** [comment]
Fixed by #1917
